### PR TITLE
Refactor yield await classification

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes parsing error allowed in test262
| Patch: Bug Fix?          | Yes
| Tests Added + Pass?      | Yes
| License                  | MIT

Currently `@babel/parser` uses parser state to track the position and possible declaration error of `yield` and `await` in ambiguous patterns: arrow head (`(x)`) and async arrow head `async(x)`. Historically we have done several fixing attempts (#6877, #7727, #9405, #10469, #11148, etc.), however this approach is error-prone and can not handle edge cases especially if these patterns are nested: For example, many, if not most, test cases I added in this PR are failing on `main` but fixed by this PR.

We introduce an `ExpressionScope` to handle declaration errors in ambiguous patterns. See [here](https://github.com/JLHwung/babel/blob/ae9acf0bfca9c568642103e912aca46949c0cadf/packages/babel-parser/src/util/expression-scope.js#L7-L45) for the design of `ExpressionScope`. The idea is to record such errors and only thrown at the time when such pattern is indeed an arrow function or an async arrow function. The stack based scope play nicely with nested structures: so errors will not be thrown too early or incorrectly reset when later a new structure is parsed.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12230"></a>

### CODE REVIEW

File: packages/babel-parser/src/parser/expression.js [hunk 1]
```diff
@@ -743,47 +742,17 @@ export default class ExpressionParser extends LValParser {
...
-
-      // We keep the old value if it isn't null, for cases like
-      //   (x = async(yield)) => {}
-      //
-      // Hi developer of the future :) If you are implementing generator
-      // arrow functions, please read the note below about "await" and
-      // verify if the same logic is needed for yield.
-      if (oldYieldPos !== -1) this.state.yieldPos = oldYieldPos;
-
-      // Await is trickier than yield. When parsing a possible arrow function
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/expression.js:765]: @nicolo-ribaudo I really love reading these informative comments.

File: packages/babel-parser/test/fixtures/es2015/yield/parameter-default-inside-arrow-inside-generator-5/output.json [hunk 2]
```diff
@@ -2,8 +2,9 @@
   "type": "File",
   "start":0,"end":47,"loc":{"start":{"line":1,"column":0},"end":{"line":3,"column":1}},
   "errors": [
-    "SyntaxError: Can not use 'yield' as identifier inside a generator (2:8)",
-    "SyntaxError: Binding invalid left-hand side in function parameter list (2:8)"
+    "SyntaxError: yield is not allowed in generator parameters (2:8)",
+    "SyntaxError: Binding invalid left-hand side in function parameter list (2:8)",
+    "SyntaxError: yield is not allowed in generator parameters (2:8)"
```
Comment (JLHwung) [inline packages/babel-parser/test/fixtures/es2015/yield/parameter-default-inside-arrow-inside-generator-5/output.json:7]: This is, in a way, the side effect of recording errors in every ancestry scope (until expression scope).

https://github.com/babel/babel/blob/d51aa6d76177b544590cdfe3868f9f4d33d8813d/packages/babel-parser/test/fixtures/es2015/yield/parameter-default-inside-arrow-inside-generator-5/input.js#L1-L3

In the error recovery mode, since the error is recorded in both `(yield)` and (x = (yield) => {})`, it will be thrown twice although they are both caused by `yield`. We could cleanup ancestry errors in `validateAsPattern` -- but we are not aware whether certain errors are caused by the same `yield` identifier`.

Note that we already throw 2 different errors for this case, which is not ideal.

File: packages/babel-parser/test/fixtures/es2015/yield/parameter-name-arrow-inside-generator-1/output.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "type": "File",
   "start":0,"end":35,"loc":{"start":{"line":1,"column":0},"end":{"line":3,"column":1}},
   "errors": [
-    "SyntaxError: Can not use 'yield' as identifier inside a generator (2:3)",
+    "SyntaxError: yield is not allowed in generator parameters (2:3)",
```
Comment (JLHwung) [inline packages/babel-parser/test/fixtures/es2015/yield/parameter-name-arrow-inside-generator-1/output.json:5]: https://github.com/babel/babel/blob/d51aa6d76177b544590cdfe3868f9f4d33d8813d/packages/babel-parser/test/fixtures/es2015/yield/parameter-default-inside-arrow-inside-generator-1/input.js#L1-L3

The error message seems a bit misleading as it is not a generator, although it is indeed in function parameters. V8 has changed this error message to "Yield expression not allowed in formal parameter" ([code](https://source.chromium.org/chromium/chromium/src/+/master:v8/src/common/message-template.h;l=543;drc=bd16dc8f9d636755d54ee27bc95435afd258552d), I suggest we align with V8 and update the error in following PRs.

File: packages/babel-parser/test/fixtures/es2017/async-functions/await-binding-inside-arrow-params-inside-async-arrow-params/output.json [hunk 4]
```diff
@@ -2,7 +2,7 @@
   "type": "File",
   "start":0,"end":36,"loc":{"start":{"line":1,"column":0},"end":{"line":1,"column":36}},
   "errors": [
-    "SyntaxError: Can not use 'await' as identifier inside an async function (1:20)"
+    "SyntaxError: Can not use 'await' as identifier inside an async function (1:14)"
```
Comment (JLHwung) [inline packages/babel-parser/test/fixtures/es2017/async-functions/await-binding-inside-arrow-params-inside-async-arrow-params/output.json:5]: https://github.com/babel/babel/blob/d51aa6d76177b544590cdfe3868f9f4d33d8813d/packages/babel-parser/test/fixtures/es2017/async-functions/await-binding-inside-arrow-params-inside-async-arrow-params/input.js#L1

The position of `await` should 14.

File: packages/babel-parser/src/util/expression-scope.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  isCertainlyParameterDeclaration() {
+    return this.type === kParameterDeclaration;
+  }
+}
+
+class ArrowHeadParsingScope extends ExpressionScope {
+  declare declarationErrorPos: number;
+  declare declarationErrorMessage: string;
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/util/expression-scope.js:77]: When in `errorRecovery` mode, there could be multiple errors. For example, in `async (x = await, y = await) => {}`.

We can thus store this as a map, and track all of them:
```suggestion
  errors: Map</* pos */ number, /* message */ string> = new Map();
```

File: packages/babel-parser/src/util/expression-scope.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    const currentScope = stack[stack.length - 1];
+    if (!currentScope.canBeArrowParameterDeclaration()) return;
+    /*:: invariant(currentScope instanceof ArrowHeadParsingScope) */
+    if (currentScope.hasDeclarationError()) {
+      this.raise(
+        currentScope.declarationErrorPos,
+        /* eslint-disable @babel/development-internal/dry-error-messages */
+        currentScope.declarationErrorMessage,
+      );
+    }
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/util/expression-scope.js:165]: To avoid duplicate errors in the ancestor scopes, we can do something like this to delete the duplciates (assuming the Map implementation):
```js
let i = stack.length - 1;
const currentScope = stack[i];

for (const [pos, message] of currentScope.errors) {
  while (i >= 0 && stack[i].errors.delete(pos)) i--;

  this.raise(pos, message);
}
```

**Review by nicolo-ribaudo — APPROVED** [review]
This is awesome :tada:

**Review by existentialism — APPROVED** [review]
